### PR TITLE
Soft NMS for Boxes and RotatedBoxes

### LINKED ISSUE #980: Soft NMS
Hello! I see soft NMS in detectron1, so will there ever be support of soft NMS in detectron2?

### PR DESCRIPTION
**Pull request by alekseynp**
Resolves #980 

A pytorch implementation of Soft NMS for both Boxes and RotatedBoxes.

I have a cyclical dependency problem if I try to import soft_nms into layers. This is driven by the fact that `pairwise_iou` expected Boxes even though it only cares about tensors. In this way soft_nms needs access to structures to get Boxes. Open to suggestions for structural changes.

I have not run eval yet on pre-trained models to verify.

Unit tests pass except `test_model_e2e.py::MaskRCNNE2ETest::test_rpn_inf_nan_data` which I think isn't my fault?

### CODE REVIEW

**Comment by alekseynp** [comment]
I ran evaluations on pretrained models from the model zoo.

**COCO Object Detection Baselines**
Faster R-CNN:
- R50-C4 1x
- X101-FPN 3x

RetinaNet:
- R101 3x

**COCO Instance Segmentation Baselines with Mask R-CNN**
- X101-FPN 3x

Base is as-is with no config changes.
Linear is soft nms enabled, mode linear, nms_thresh_test 0.3
Gaussian is soft_nms_enabled, mode gaussian, sigma 0.5

In all cases I think we see the kind of improvements expected.
The implementation is a little slow, but it directly leverages the existing pairwise_iou functions which is especially useful for rotation.
The plumbing is all there if better cpu/gpu axis aligned/rotated implementations were to be integrated.

```
MODEL ZOO
R50-C4 1x
box AP: 35.7

BOX
|   nms    |   AP   |  AP50  |  AP75  |  APs   |  APm   |  APl   |
|:--------:|:------:|:------:|:------:|:------:|:------:|:------:|
| base     | 35.681 | 56.098 | 38.027 | 19.234 | 40.860 | 48.712 |
| linear   | 36.400 | 56.632 | 39.379 | 19.519 | 41.471 | 49.763 |
| gaussian | 36.506 | 56.504 | 39.617 | 19.616 | 41.593 | 49.910 |

MODEL ZOO
X101-FPN 3x
box AP: 43.0

BOX
|   nms    |   AP   |  AP50  |  AP75  |  APs   |  APm   |  APl   |
|:--------:|:------:|:------:|:------:|:------:|:------:|:------:|
| base     | 43.047 | 63.652 | 46.888 | 27.205 | 46.093 | 54.890 |
| linear   | 43.908 | 63.858 | 48.376 | 27.778 | 46.979 | 55.699 |
| gaussian | 43.762 | 63.375 | 48.234 | 27.714 | 46.730 | 55.593 |

MODEL ZOO
RetinaNet R101 3x
box AP: 39.9

BOX
|   nms    |   AP   |  AP50  |  AP75  |  APs   |  APm   |  APl   |
|:--------:|:------:|:------:|:------:|:------:|:------:|:------:|
| base     | 39.874 | 60.135 | 42.795 | 23.841 | 43.531 | 51.719 |
| linear   | 40.130 | 60.312 | 43.501 | 23.783 | 43.569 | 52.160 |
| gaussian | 40.305 | 60.274 | 43.724 | 23.909 | 43.704 | 52.422 |

MODEL ZOO
Mask R-CNN
X101-FPN 3x
box AP: 44.3
mask AP: 39.5

BOX
|   nms    |   AP   |  AP50  |  AP75  |  APs   |  APm   |  APl   |
|:--------:|:------:|:------:|:------:|:------:|:------:|:------:|
| base     | 44.275 | 64.463 | 48.618 | 27.530 | 47.637 | 56.698 |
| linear   | 45.178 | 64.730 | 50.162 | 28.103 | 48.584 | 57.681 |
| gaussian | 45.009 | 64.251 | 49.994 | 28.018 | 48.359 | 57.590 |

MASK
|   nms    |   AP   |  AP50  |  AP75  |  APs   |  APm   |  APl   |
|:--------:|:------:|:------:|:------:|:------:|:------:|:------:|
| base     | 39.520 | 61.696 | 42.570 | 20.680 | 41.967 | 56.520 |
| linear   | 39.920 | 62.008 | 43.181 | 20.901 | 42.376 | 57.073 |
| gaussian | 39.713 | 61.562 | 42.953 | 20.786 | 42.237 | 56.866 |
```

**Comment by ashnair1** [comment]
Any updates on this?

**Comment by dcela** [comment]
What is missing to get this merged (besides the conflicts)?

**Comment by alekseynp** [comment]
The maintainers have not shown particular interest in this PR. I originally wrote this work because I wanted to use Soft NMS and it worked great. Later I discovered the dark and mysterious `torch.ops._caffe2.BoxWithNMSLimit` ([docs](https://caffe2.ai/docs/operators-catalogue.html#boxwithnmslimit)) Who knows what goes on in private between those who work at FB on PT, Caffe2, detectron2, etc. It feels gross and dangerous to use the `_` op, but on the other hand, when I `torch.jit.trace` I am able to transfer to C++ seamlessly which is nice. If you search this repo for `torch.ops._caffe2` you will find some more like it. This code is probably just an `@torch.jit.script` decorator with maybe a little type hinting away from being traceable, but the cpp code in the `_caffe2` op is probably faster. Sorry @ashnair1 @dcela for the delayed comment. I wouldn't count on this ever being merged, but I could conceivably update this PR if you wanted to merge it into your own fork.

**Comment by dcela** [comment]
No worries, glad there is an alternative! And now if someone searches for "soft NMS detectron2" they can hopefully find this

**Comment by Syzygianinfern0** [comment]
@dcela How can I use this in my code? Either the code in the PR or the one from the caffe docs. I would really like to use `SoftNMS` in my code. 

Is there anything more straightforward cause 
- merging this PR into my local master branch
- resolving conflicts
- building detectron2 from source 

seems to be the only thing that I think that can be done.

**Comment by longyueying** [comment]
@alekseynp 
I implemented soft-nms  following your code。AP equal to 35.7 when setted method as hard  , which was same as original nms. But AP was lower than original nms when I setted method as linear or gassian。The AP results as follows when used linear  soft-nms :
BOX
|   nms    |   AP   |  AP50  |  AP75  |  APs   |  APm   |  APl   |  nms threshold   |  score threshold|
|:--------:|:------:|:------:|:------:|:------:|:------:|:------:|:------:|:------:|
| base     | 35.681 | 56.098 | 38.027 | 19.234 | 40.860 | 48.712 |
| linear   | 32.3458 | 48.4509 | 35.8470 | 17.9183 | 37.8233 | 44.7475 |0.3|0.05|
| linear   | 33.5307 | 50.6488 | 37.0284 | 17.8386 | 39.2859 | 46.1550 |0.3|0.1|
| linear   | 34.2221 | 52.3293 | 37.4397 | 17.6601 | 39.7792 | 47.0231 |0.3|0.2|
| linear   | 34.0554 | 52.3908 | 39.9727 | 16.9642 | 39.2221 | 47.2343 |0.3|0.3|

I have checked the implementation over and over，but no error was found.
The soft-nms was implemented as follows，could you help me to figure the problem out. 
thx a lot!

```python
import torch
from detectron2.structures import Boxes, pairwise_iou
                                                                                                                                                                                       
def soft_nms(boxes, scores, method, Nt, sigma, threshold):
    out = []
           
    boxes = boxes.clone()
    scores = scores.clone()
    idxs = torch.arange(scores.size()[0])

    while scores.numel() > 0:
        top_idx = torch.argmax(scores)
        top_box = boxes[top_idx]
        ious = pairwise_iou(Boxes(top_box.unsqueeze(0)), Boxes(boxes))[0]

        if method == 'linear':
            decay = torch.ones_like(ious)
            decay_mask = ious > Nt
            decay[decay_mask] = 1 - ious[decay_mask]
        elif method == 'gaussian':
            decay = torch.exp(-torch.pow(ious, 2) / sigma)
        elif method == "hard":
            decay = (ious < Nt).float()
        else:
            raise NotImplementedError("{} Not Implemented Yet".format(method))

        scores *= decay
        keep = scores > threshold
        keep[top_idx] = False

        out.append(idxs[top_idx])
        boxes = boxes[keep]
        scores = scores[keep]
        idxs = idxs[keep]

    return torch.tensor(out)


def batched_soft_nms(boxes, scores, idxs, method='linear', Nt=0.3, sigma=0.5, threshold=0.2):
    if boxes.numel() == 0:
        return torch.empty((0), dtype=torch.int64, device=boxes.device)
    max_coordinate = boxes.max()
    offsets = idxs.to(boxes) * (max_coordinate + 1)
    boxes_for_nms = boxes + offsets[:, None]
    keep = soft_nms(boxes_for_nms, scores, method, Nt, sigma, threshold)
    return keep
```

**Comment by alekseynp** [comment]
I haven't had the time to consider this carefully, but I do note that you have merged the `linear_threshold` and `prune_threshold` from my implementation into one.

**Comment by longyueying** [comment]
@alekseynp 
After you mentioned **linear_threshold** and **prune_threshold**, I realize that re-score for soft-nms was missed. Expected result has gotted after adding re-score for soft-nms was missed. thanks a lot!
